src/protocol/sender.rs

```diff
@@ -0,0 +1,622 @@
+use crate::loss_compression::decompress_loss_list;
+use crate::packet::{ControlTypes, HandshakeControlInfo, PacketLocation, SrtControlPacket};
+use crate::protocol::{Handshake, TimeBase, TimeStamp};
+use crate::Packet::*;
+use crate::{
+    CCData, CongestCtrl, ConnectionSettings, ControlPacket, DataPacket, MsgNumber, Packet,
+    SeqNumber, SocketID, SrtCongestCtrl,
+};
+use bytes::Bytes;
+use std::collections::VecDeque;
+use std::net::SocketAddr;
+use std::time::{Duration, Instant};
+
+pub struct TransmitBuffer {
+    remote_sockid: SocketID,
+    max_packet_size: usize,
+    time_base: TimeBase,
+
+    /// The list of packets to transmit
+    buffer: VecDeque<DataPacket>,
+
+    /// The sequence number for the next data packet
+    next_seq_number: SeqNumber,
+
+    /// The message number for the next message
+    next_message_number: MsgNumber,
+}
+
+impl TransmitBuffer {
+    pub fn new(settings: &ConnectionSettings) -> Self {
+        Self {
+            remote_sockid: settings.remote_sockid,
+            max_packet_size: settings.max_packet_size as usize,
+            time_base: TimeBase::from_raw(settings.socket_start_time),
+            buffer: Default::default(),
+            next_seq_number: settings.init_seq_num,
+            next_message_number: MsgNumber::new_truncate(0),
+        }
+    }
+
+    /// In the case of a message longer than the packet size,
+    /// It will be split into multiple packets
+    pub fn push_data(&mut self, data: (Bytes, Instant)) {
+        let (mut payload, time) = data;
+        let mut location = PacketLocation::FIRST | PacketLocation::LAST;
+        while !payload.is_empty() {
+            // if we need to break this packet up
+            if payload.len() > self.max_packet_size {
+                let slice = payload.slice(self.max_packet_size..payload.len());
+
+                self.begin_transmit(time, slice, location);
```

Comment:
The `LAST` flag will get set on the first one. I think the logic in this function needs some close thinking. Maybe a unit test would help.